Drop down selection box

shusain opened the pull request.

shusain committed:
Adds some starting point for the drop down selection directive.

shusain committed:
Just ignoring IntelliJ config files.

shusain committed:
Adds font-awesome for the sake of getting some okay looking chevrons generally available.

shusain committed:
Updates to get drop downs working in a basic way at least.

Still needs some finishing up with regard to keyboard event handling and
customization of list items to allow elements in the list that are not
selectable and for handling default selection properly.  Does the basics
of open up allow a user to click an element, has custom templates for
the list items and a template for the placeholder/chosen value display.
Intelligently opens up or down depending on the position of the
element on screen, if more than half way down opens upward and reverses
the list items, otherwise opens downwards.

shusain committed:
Just noticed the last couple of updates weren't included in the built version in the dist folder so built that gain.

hassanbazzi commented:
font-awesome is an enormous library. We should be doing all this CSS / styling things inside main-app instead

shusain committed:
Updates that get everything functional

Have the keyboard interactivity and the non interactive items and list
ordering all working better now.  Just doing a commit before cleaning
things up to make this easy to style and customize per application.

shusain committed:
Removed font-awesome from dependencies, updated some of the SASS

shusain committed:
Put a few things in watches in case the data comes later, added a callback option for the quick action drop downs.

hassanbazzi commented on src/dropDown.js:
```diff
@@ -0,0 +1,237 @@
+          if(middleOfElement>middleOfWindow){
+            $scope.direction = 'up';
+
+            dropDownListContainer[0].style.bottom = rect.height+'px';
```
`bottom: 100%` achieves the same effect without calculation

hassanbazzi commented:
Conflicts. Honestly, there's a ton of code in this. I feel it's a bit over-engineered for a dropdown menu but all the functionality is there. LGTM once conflicts are resolved.

shusain commented:
Agree in general I wish it were less code and it needs to be better organized so it can be properly cleaned up.  That said I don't really know anything off hand that's in here that wasn't added out of necessity to keep all the current functionality.  Only bit of added functionality is for the keyboard support but it is not really a substantial part of the code.

On the plus side this should clean up a bit of the markup in the actual app since we'll just be defining the list element templates but the rest of the control is tucked away behind the directive.

Just saw there's an issue with the tooltips not cleaning up when switching states so going to get that fixed in another PR but will fix conflicts here first.

shusain commented:
Just a note seems most of those conflicts were just because of rebuilding the dist folder every time, maybe good to figure out a plan for that so it's not always in conflict if two people are working or someone is working on more than one branch in here.

shusain commented:
@hassanbazzi Okay to merge this?

hassanbazzi commented:
LGTM
